Initial CLI Version of `cfssl scan`

jacobhaven opened the pull request:
Adds new `scan` command, that performs a variety of connectivity and TLS protocol tests.

Unfortunately, some of these require importing the go standard library's `crypto/tls` package's source in order to access private members. These are used in `scan/tls/cfsslscan_*.go` files to export additional functionality not supplied by the `crypto/tls` API.

Support for the standard CFSSL HTTP API is forthcoming, but comments on this initial structure might be useful first.

lziest commented on scan/scan_common.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+// Scan defines the scan to be performed on the given host.
+func (s *Scanner) Scan(host string) (grade Grade, output Output, err error) {
```
Looks good. One style issue: For exported function, we keep the return list nameless, so the doc will be consistent with standard GoDoc. Also, it makes each return statement declare exactly what should be returned. It's much easier to maintain afterwards, and reduces potential bugs.

jacobhaven replied:
Good point, done.

lziest commented on scan/tls_ciphersuite_scanner.go:
```diff
@@ -0,0 +1,65 @@
+	"errors"
+	"net"
+
+	"github.com/jacobhaven/cipherscan/tls"
```
Some compilation error regarding this package.

grittygrease replied:
Don't we want the version from cfssl/scan/tls here?

jacobhaven replied:
Ah, well that would explain something. \*updating\*

lziest commented on scan/tls_session_scanner.go:
```diff
@@ -0,0 +1,63 @@
+	config := &tls.Config{ClientSessionCache: sessions}
+	for _, addr := range addrs {
+		var conn *tls.Conn
+		conn, err = tls.Dial(Network, addr, config)
```
I recommend to use DialWithDialer, so we can control the timeout and easy to tweak. It's otherwise the system default timeout, which could be like 30 seconds.

grittygrease commented on cfssl_scan.go:
```diff
@@ -0,0 +1,82 @@
+
+Usage of scan:
+        cfssl scan [-short] [SCANFAMILY] HOST
+        cfssl scan [-short] [-scanner regexp] HOST
```
Please put an example here. The documentation is confusing, is SCANFAMILY a regexp? What is regexp matching against? What does [SCANFAMILY]* represent?

jacobhaven replied:
I've refactored this to be a normal argument `-scanfamily regexp`

grittygrease commented on cfssl.go:
```diff
@@ -67,6 +67,9 @@ var Config struct {
 	remote            string
+	short             bool
+	list              bool
+	scanner           string
```
Need to register flags and command (CLIScan) below.

kisom commented on scan/connectivity_scanner.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+// DNSLookupScanner tests that DNS resolution of the host returns at least one address
+var DNSLookupScanner = &Scanner{
```
Should use the field names in initialisation.

kisom commented on scan/connectivity_scanner.go:
```diff
@@ -0,0 +1,57 @@
+
+// ConnectivityScans contains scanners testing basic connectivity to the host
+var ConnectivityScanners = &Family{
+	Name:        "ConnectivityScanners",
```
Style note: if Family denotes a family of scanners, you can probably drop the "Scanners" part of the name. if this is the name that the user selects the family with on the commandline, I would just use "connectivity" to make it easier and consistent with the other args.

kisom commented on scan/connectivity_scanner.go:
```diff
@@ -0,0 +1,57 @@
+		output = []string{}
+		return
+	}
+	defer conn.Close()
```
This shouldn't be deferred; the conn is no longer in use and should be closed right away.

jacobhaven replied:
I didn't , this was a a relic of before this test was split from another one (It actually caused a bug and was removed)

kisom commented on scan/scan_common.go:
```diff
@@ -0,0 +1,78 @@
+// Output is the result of a Scan, to be stored for potential use by later Scanners.
+type Output interface{}
+
+// Scanner describes a type of scan to perform on a host.
```
One thing I'm trying to move to in CFSSL is to make sure structure fields are documented as well.

jacobhaven replied:
Done.

kisom commented on scan/scan_common.go:
```diff
@@ -0,0 +1,78 @@
+}
+
+// Family defines a set of related scans meant to be run together in sequence
+type Family struct {
```
These field comments are good. As a Go style note though, they should be complete sentences.

grittygrease commented on cfssl_scan.go:
```diff
@@ -0,0 +1,82 @@
+
+Arguments:
+        HOST:    Host to scan (including port)
+        SCANFAMILY:	 Regexp selecting family of related scanners to run
```
Inconsistency with spaces v. tabs.

grittygrease commented on cfssl_scan.go:
```diff
@@ -0,0 +1,82 @@
+		if Config.scanner != "" {
+			scannerRegexp = regexp.MustCompile(Config.scanner)
+		}
+		for len(args) > 1 {
```
This logic needs some work. calling cfssl scan with no arguments causes scanFamily.Name to be nil and 
			matches := scanFamilyRegexp.MatchString(scanFamily.Name)
to cause a panic.

grittygrease commented on scan/tls/alert.go:
```diff
@@ -0,0 +1,79 @@
+// Copyright 2009 The Go Authors. All rights reserved.
```
It would be very helpful to have some documentation here. At the minimum we need a revision of the original code this was copied from and a diff to see which changes are custom.

grittygrease replied:
It might even make sense to leave this as its own repository (forked from crypto/tls) so that we can track the diffs and keep it up to date with Go's tls changes.

jacobhaven replied:
This isn't a bad idea. To set it up as a submodule. I don't think we'll need to make any changes to the original `crypto/tls` files. I've just needed to access some functionality in creating the helper functions in `tls/cfsslscan_*`

grittygrease replied:
No need to use a submodule in Go, you can just include from Github. For safety, the module name could be changed to scan-tls, cf-tls or something like that.

kisom commented:
You've checked in the CFSSL binary. Please remove this (and make sure the commits are squashed so that it doesn't appear in any commits).

kisom commented on scan/tls_handshake_scanners.go:
```diff
@@ -0,0 +1,65 @@
+	serverCipher, serverVersion, err = conn.SayHello()
+	for _, cipherID := range ciphers {
+		if serverCipher == cipherID {
+			err = errors.New("Server negotiated ciphersuite we didn't send.")
```
As a style note, Go error strings are generally not capitalised and don't end in punctuation.

jgrahamc commented on cfssl_scan.go:
```diff
@@ -0,0 +1,103 @@
+	}
+	return
+}
+func scanMain(args []string) (err error) {
```
You've mixed the user of a named return argument and returns with arguments. Choose one or the other. I'd suggest making the returned value just 'error' and using explicit 'return foo'.

jacobhaven replied:
Most of the other cfssl Mains use `(err error)`, so I'll stick with that, but fix the `return err`/`return nil` instances

jgrahamc commented on cfssl_scan.go:
```diff
@@ -0,0 +1,103 @@
+var scanFlags = []string{"short", "list", "family", "scanner"}
+var scanFamilyRegexp, scannerRegexp *regexp.Regexp
+
+func regexpLoop(familyFunc func(*scan.Family) error, scannerFunc func(*scan.Scanner) error) (err error) {
```
I think its worth documenting these arguments as they are rather complex and it's not immediately clear what's expected of a familyFunc or a scannerFunc.

jacobhaven replied:
Updated with:
```
// regexpLoop iterates through each scan Family and Scanner registered in scan.AllFamilies.
// familyFunc is run on each Family with a name matching the family flag's regexp,
// then scannerFunc is run on each Scanner in that Family that matches the scanner flag's regexp
```

jgrahamc commented on scan/connectivity.go:
```diff
@@ -0,0 +1,52 @@
+		return
+	}
+	addrs, err := net.LookupHost(host)
+	if err != nil || len(addrs) == 0 {
```
If err == nil but len(addrs) == 0 this will return nil. I think you want something else.

jacobhaven replied:
Good point. Split into two separate errors.

jgrahamc commented on cfssl.go:
```diff
@@ -94,6 +98,10 @@ func registerFlags() {
 	cfsslFlagSet.StringVar(&Config.domain, "domain", "", "remote server domain name")
 	cfsslFlagSet.StringVar(&Config.ip, "ip", "", "remote server ip")
 	cfsslFlagSet.StringVar(&Config.remote, "remote", "", "remote CFSSL server")
+	cfsslFlagSet.BoolVar(&Config.short, "short", false, "so not run long/intensive scans")
```
Not sure what 'so no run long/intensive scans' means.

jacobhaven replied:
It was intended as "don't run long and/or intensive scans". But, preferably we won't have those, or you can select to not run the entire suite. The `short` argument has been removed, and and replaced with a `verbose` output argument.

jgrahamc commented on cfssl.go:
```diff
@@ -67,6 +67,10 @@ var Config struct {
 	remote            string
+	short             bool
+	list              bool
+	family            string
```
These need documenting. It's very unclear how family and scanner interact.

jacobhaven replied:
Essentially you should probably only use one or the other. Using both takes the intersection. How does this sound?
```
  -family="": regular expression selecting specific scan families to run (defaults to all)
  -scanner="": regular expression selecting specific scanners to run (defaults to all)
```

jgrahamc commented on scan/scan_common.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+type scannerHostPair struct{ scannerName, host string }
+type scanHistory struct {
```
If all you have in scanHistory is this map then just do
```
type scanHistory map[scannerHostPair][]Output
```
and simplify your life.

jacobhaven replied:
True. At first I thought there might be more included in there. I've refactored a large part of this.

jgrahamc commented on scan/tls_handshake.go:
```diff
@@ -0,0 +1,109 @@
+	}
+	return list
+}
+func (ciphers cipherList) Append(cipher uint16) {
```
This function seems a bit pointless.

jacobhaven replied:
Fair :)

jgrahamc commented on scan/tls_session.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+// sessionCache is a mock ClientSessionCache used to record all stores (i.e. new session tickets)
+type sessionCache struct {
```
This could just be 
```
type sessionCache map[string][]*tls.ClientSessionState
```

jacobhaven replied:
I've actually removed this type entirely and used `tls.NewLRUClientSessionCache()` instead

jgrahamc commented on scan/tls_session.go:
```diff
@@ -0,0 +1,87 @@
+	cache map[string][]*tls.ClientSessionState
+}
+
+func newSessionCache() *sessionCache {
```
If you make the change above then this function is unnecessary and you can just do 
```
make(sessionCache)
```
whenever one is needed.

jgrahamc commented on scan/tls_session.go:
```diff
@@ -0,0 +1,87 @@
+	s.cache[sessionKey] = append(s.cache[sessionKey], cs)
+}
+
+// SessionResumeScan tests that host is able to resume sessions accross all addresses.
```
across

jgrahamc commented on scan/tls_session.go:
```diff
@@ -0,0 +1,87 @@
+	}
+	sessions := newSessionCache()
+	output = sessions
+	config := &tls.Config{ClientSessionCache: sessions, InsecureSkipVerify: true}
```
Why 
```
InsecureSkipVerify: true
```
? I think that needs documenting.

jacobhaven replied:
I suppose there's no reason not to verify here, so I will.

jgrahamc commented on scan/tls_session.go:
```diff
@@ -0,0 +1,87 @@
+		} else {
+			addr = "[" + ip.String() + "]"
+		}
+		if len(port) == 0 {
```
Why not use
```
port == ""
```
instead of
```
len(port) == 0
```

jacobhaven replied:
Actually, port should *always* be populated (already defaulted to 443 if empty), so this is completely unnecessary.

jgrahamc commented on scan/tls_session.go:
```diff
@@ -0,0 +1,87 @@
+		}
+		fmt.Println()
+		var conn *tls.Conn
+		conn, err = tls.Dial(Network, addr+port, config)
```
There's actual a function called
```
net.JoinHostPort
```
http://golang.org/pkg/net/#JoinHostPort

jgrahamc commented on scan/tls_session.go:
```diff
@@ -0,0 +1,87 @@
+		}
+		conn.Close()
+		if len(sessions.cache) != 1 {
+			return
```
Is it OK that the error returned here is nil?

jacobhaven commented:
Rebased on new `cmd` and `cli` structure and cleaned up commit log, added new scan families for secure HTTP headers and (extending the scope of the tool), and scanner to scrape CIDR IP ranges for new Intermediate CAs to update [cfssl_trust](https://github.com/cloudflare/cfssl_trust) for Bundler.

grittygrease commented:
Unfortunately, some of the go code copied from the standard library does not follow the official style guidelines and fails go vet. Please check the travis output and fix.

grittygrease commented on scan/https.go:
```diff
@@ -0,0 +1,86 @@
+	return
+}
+
+// hstsScan tests that the host serves correct HPKP headers.
```
Do you also want to check for expiry time here? Short period HSTS is less valuable than long period HSTS.

jacobhaven replied:
Possibly. I mainly couldn't find an easy way to parse the header values.

grittygrease commented:
@jacobhaven Can you add tests for these packages? Even a nearly empty test file in each package makes sure that the overall coverage takes this code into account.

lziest commented on scan/scan_common.go:
```diff
@@ -0,0 +1,142 @@
+	Good
+)
+
+func (g Grade) String() string {
```
Minor, add doc.

lziest commented on scan/scan_common.go:
```diff
@@ -0,0 +1,142 @@
+	History.store(s.Name, host, output)
+	return grade, output, err
+}
+func (s *Scanner) String() string {
```
Minor, add doc and a blank line

lziest commented on scan/scan_common.go:
```diff
@@ -0,0 +1,142 @@
+	Scanners []*Scanner
+}
+
+func (f *Family) String() string {
```
Minor, add doc.

lziest commented on scan/tls_handshake.go:
```diff
@@ -0,0 +1,111 @@
+	err = fmt.Errorf("server negotiated ciphersuite we didn't send: %s", tls.CipherSuites[serverCipher])
+	return
+}
+func allCiphersIDs() []uint16 {
```
Add blank lines for style consistency.

lziest commented on cli/scan/scan.go:
```diff
@@ -0,0 +1,121 @@
+	w.Flush()
+}
+func scanMain(args []string, c cli.Config) (err error) {
+	familyRegexp := regexp.MustCompile(c.Family)
```
Minor issue: do we want to panic here if the input is not valid regex? Just use Compile and throw error here?

kisom replied:
I think that we should use Compile,m and return an error instead of panicking.

lziest commented on cli/cli.go:
```diff
@@ -94,6 +97,9 @@ func registerFlags(c *Config, f *flag.FlagSet) {
 	f.StringVar(&c.Remote, "remote", "", "remote CFSSL server")
 	f.StringVar(&c.Label, "label", "", "key label to use in remote CFSSL server")
+	f.BoolVar(&c.List, "list", false, "list possible scanners")
+	f.StringVar(&c.Family, "family", "", "family regular expression")
```
I prefer 'scanner family regular expression'

lziest commented:
Other than those minor issues, LGTM

grittygrease commented:
Once Zi's comments are addressed and the _test.go files are added, I think this is ready to merge.

kisom commented on scan/connectivity.go:
```diff
@@ -0,0 +1,80 @@
+	host, _, err = net.SplitHostPort(host)
+	if err != nil {
+		return
+	}
```
It's a style nit, but I find the code is more readable if you put a blank line here (and L53).

kisom commented on scan/https.go:
```diff
@@ -0,0 +1,86 @@
+	if err != nil {
+		return
+	}
+	hpkpHeader := resp.Header.Get("Public-Key-Pins")
```
Style nit: I think it's more readable if you put a newline between the '}' and this line.

kisom commented on scan/tls_handshake.go:
```diff
@@ -0,0 +1,111 @@
+	if err != nil {
+		return
+	}
+	if serverVersion != vers {
```
I'd add a blank line here.

kisom commented on scan/tls_handshake.go:
```diff
@@ -0,0 +1,111 @@
+		err = fmt.Errorf("server negotiated protocol version we didn't send: %s", tls.Versions[serverVersion])
+		return
+	}
+	var cipherID uint16
```
and probably here, too.

kisom commented on scan/tls_handshake.go:
```diff
@@ -0,0 +1,111 @@
+	cipherID uint16
+	versions []uint16
+}
+type cipherVersionList []cipherVersions
```
Should add a blank line.

kisom commented on scan/tls_handshake.go:
```diff
@@ -0,0 +1,111 @@
+	}
+	return strings.Join(cvStrings, "\n")
+}
+func (cvList cipherVersionList) Describe() string {
```
Blank line above this.

kisom commented on scan/tls_session.go:
```diff
@@ -0,0 +1,55 @@
+		return
+	}
+	conn.Close()
+	for _, ip := range ips {
```
I'd add a blank line about this one (after the `conn.Close()`).

kisom commented:
Only nits I had were stylistic; other than those, LGTM.

kisom commented:
Also, @jacobhaven you'll need to resolve the merge conflicts before I can merge this.

grittygrease commented:
@jacobhaven Can you address Kyle's comments so we can merge and move on to adding an API?

lziest commented on cli/config.go:
```diff
@@ -72,6 +75,9 @@ func registerFlags(c *Config, f *flag.FlagSet) {
 	f.StringVar(&c.RevokedAt, "revoked-at", "now", "Date of revocation (YYYY-MM-DD)")
 	f.Int64Var(&c.Interval, "interval", int64(4*helpers.OneDay), "Interval between OCSP updates, in seconds (default: 4 days)")
+	f.BoolVar(&c.List, "list", false, "list possible scanners")
+	f.StringVar(&c.Family, "family", "", "family regular expression")
```
Can we use "scanner family regular expression" for the flag usage string?
